src/ray/core_worker/transport/direct_task_transport.cc

```diff
@@ -75,12 +75,33 @@ Status CoreWorkerDirectTaskSubmitter::SubmitTask(TaskSpecification task_spec) {
             task_spec.GetSchedulingClass(), task_spec.GetDependencyIds(),
             task_spec.IsActorCreationTask() ? task_spec.ActorCreationId()
                                             : ActorID::Nil());
-        auto it = task_queues_.find(scheduling_key);
-        if (it == task_queues_.end()) {
-          it =
-              task_queues_.emplace(scheduling_key, std::deque<TaskSpecification>()).first;
+        auto &scheduling_key_entry = scheduling_key_entries_[scheduling_key];
+        scheduling_key_entry.task_queue.push_back(task_spec);
+        if (scheduling_key_entry.total_tasks_in_flight <
+            scheduling_key_entry.active_workers.size() *
+                max_tasks_in_flight_per_worker_) {
```

Comment:
Suggest wrapping this in a `const` method of the `SchedulingKeyEntry` (e.g., `HasAvailableWorkers()`) to make it more readable!

Reply:
I was wondering, what do you mean by a `const` method? Also, I needed to define the method outside of the `SchedulingKeyEntry` struct to be able to access the `max_tasks_in_flight_per_worker_` member of `CoreWorkerDirectTaskSubmitter`. Otherwise, the compiler reported an error. Do you know if there is a way to access `max_tasks_in_flight_per_worker_` from a method within the `SchedulingKeyEntry` struct?

Reply:
Ah I just mean to mark that the method does not modify the `SchedulingKeyEntry`, like this:
`CanDelete() const { ... }`

Yes, it shouldn't be able to access it because it's a private member of `CoreWorkerDirectTaskSubmitter`. You can fix it by allowing `CanDelete` to take the max tasks as an argument.

Reply:
Sounds good! I'll add the const keyword, and allow CanDelete() (and the other two methods I added to check whether the pipeline fullness) to take `max_tasks_in_flight_per_worker_` as an argument, so that I can place the function inside the struct